src/wheel/wheelfile.py

```diff
@@ -1,169 +1,311 @@
-
-        ef_name = as_unicode(name_or_info.filename if isinstance(name_or_info, ZipInfo)
-                             else name_or_info)
-        if mode == 'r' and not ef_name.endswith('/') and ef_name not in self._file_hashes:
-            raise WheelError("No hash found for file '{}'".format(native(ef_name)))
-
-        ef = ZipFile.open(self, name_or_info, mode, pwd)
-        if mode == 'r' and not ef_name.endswith('/'):
-            algorithm, expected_hash = self._file_hashes[ef_name]
-            if expected_hash is not None:
-                # Monkey patch the _update_crc method to also check for the hash from RECORD
-                running_hash = hashlib.new(algorithm)
-                update_crc_orig, ef._update_crc = ef._update_crc, _update_crc
-
-        return ef
-
-    def write_files(self, base_dir):
-        logger.info("creating '%s' and adding '%s' to it", self.filename, base_dir)
+                self._zip.close()
+            finally:
+                self._record_entries.clear()
+
+    def __enter__(self) -> 'WheelFile':
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    @staticmethod
+    def _get_zipinfo_datetime(timestamp: float) -> Tuple[int, int, int, int, int, int]:
+        # Some applications need reproducible .whl files, but they can't do this without forcing
+        # the timestamp of the individual ZipInfo objects. See issue #143.
+        timestamp = int(os.environ.get('SOURCE_DATE_EPOCH', timestamp))
+        return time.gmtime(timestamp)[0:6]
+
+    def write_file(self, archive_name: str, contents: Union[bytes, str, PathLike]) -> None:
+        timestamp = time.time()
+        if isinstance(contents, bytes):
+            data = contents
+        elif isinstance(contents, str):
+            data = contents.encode('utf-8')
+        elif isinstance(contents, PathLike):
+            path = Path(contents)
+            timestamp = path.stat().st_mtime
+            data = path.read_bytes()
+        else:
+            raise TypeError('contents must be a str, bytes or a path-like object')
+
+        if archive_name not in self._exclude_archive_names:
+            hash_digest = hashlib.new(self._default_hash_algorithm, data).digest()
+            self._record_entries[archive_name] = WheelRecordEntry(
+                self._default_hash_algorithm, hash_digest, len(data))
+
+        zinfo = ZipInfo(archive_name, date_time=self._get_zipinfo_datetime(timestamp))
+        zinfo.external_attr = 0o664 << 16
+        self._zip.writestr(zinfo, data)
+
+    def write_data_file(self, archive_name: str, contents: Union[bytes, str, PathLike]) -> None:
+        archive_path = self._data_path + '/' + archive_name
+        self.write_file(archive_path, contents)
+
+    def write_metadata_file(self, archive_name: str,
+                            contents: Union[bytes, str, PathLike]) -> None:
+        archive_path = self._dist_info_path + '/' + archive_name
```

Comment:
Consider using posixpath.join, for these?

Reply:
Why?

Reply:
Agreed. `posixpath.join` would mean that `write_metadata_file("/absolute/path", "")` would do the wrong thing. String concatenation would give an invalid filename (I don't know how the zipfile module would handle it) but at least it's not silently adding an absolute path.

Reply:
Ah, my question was prompted by a misunderstanding. Sure, I can do that.

Reply:
I'm not sure my comment read correctly.

You've now changed to use `posixpath.join`. But with that change, `w.write_metadata_file("/boom", "Dummy data")` will write a wheel file that contains the file "/boom", which is an absolute path that will be installed outside of site-packages. That feels like a pretty bad error.

I was arguing that `posixpath.join` was the *wrong* thing to use here. Did *I* misunderstand something?

Reply:
Github did not display your comment before I wrote mine.

Reply:
@pradyunsg what was your original reasoning for using `posixpath.join()` here?

Reply:
At the very least I should probably strip any leading slashes from `archive_name`.

Reply:
I've reverted the change for now.

Reply:
> Github did not display your comment before I wrote mine.

Ah, I wondered if it might be that - the timestamps looked a little odd to me, too.